Improve handling of SVG images with subresources

annevk opened the pull request:
#### 84cd36c4d819ad38ac12b3b95fd274b68876ec36
<pre>
Improve handling of SVG images with subresources
<a href="https://bugs.webkit.org/show_bug.cgi?id=290840">https://bugs.webkit.org/show_bug.cgi?id=290840</a>

Reviewed by Youenn Fablet.

When a resource is completely fetched, ImageLoader::notifyFinished is
called. This does not quite work for SVG images as they can have local
subresources (data: URLs in particular).

In 292944@main we introduced some infrastructure to handle this better
for ImageUtilities callers. Notably that did not handle multiple layers
of nesting, because it did not improve the situation for HTML &lt;img&gt; and
SVG &lt;image&gt; elements.

In particular if you have a document _DOC_ that references an SVG image
_IMAGE_. And _IMAGE_&apos;s document has an SVG &lt;image&gt; that references
another SVG image _IMAGE2_. And _IMAGE2_&apos;s document references another
image _IMAGE3_, then _IMAGE_&apos;s subresourcesAreFinished() would return
before _IMAGE3_ is ready, because _IMAGE2_ did not increase the delay of
_IMAGE_&apos;s document&apos;s load event.

So now we add the infrastructure for those elements to ImageLoader and
SVGImage.

The final bit of complexity here is around supporting multiple
invocations of SVGImage::subresourcesAreFinished by improving
Document::whenWindowLoadEventOrDestroyed. This is needed because
multiple ImageLoader instances can end up pointing to the same SVGImage
instance. E.g., in our test coverage this is hit through SVG &lt;use&gt;
usage.

* LayoutTests/TestExpectations:
* LayoutTests/svg/as-image/svg-image-with-data-uri-from-canvas.html:
* LayoutTests/svg/as-image/svg-image-with-data-uri-reloading.html:
* LayoutTests/svg/as-image/svg-image-with-svg-data-uri.html:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::whenWindowLoadEventOrDestroyed):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::notifyFinished):
(WebCore::ImageLoader::setImageCompleteAndMaybeUpdateRenderer):
* Source/WebCore/loader/ImageLoader.h:
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::subresourcesAreFinished):
* Source/WebCore/platform/graphics/Image.h:
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::subresourcesAreFinished):
(WebCore::SVGImage::tryCreateFromData):
* Source/WebCore/svg/graphics/SVGImage.h:

Canonical link: <a href="https://commits.webkit.org/293259@main">https://commits.webkit.org/293259@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/400f68ec0a3e92463173c51cd65a1b4e7d12a112

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98304 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17935 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8163 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103422 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48833 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100349 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18227 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26386 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74822 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31990 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101308 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13799 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88782 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55182 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13582 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6742 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48275 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83551 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6816 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105797 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25391 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18479 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83813 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25764 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84984 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83278 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27915 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5590 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19030 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15935 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25349 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30523 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25169 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28485 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26744 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -486,18 +487,32 @@ void ImageLoader::notifyFinished(CachedResource& resource, const NetworkLoadMetr
-        decode();
-    loadEventSender().dispatchEventSoon(*this, eventNames().loadEvent);
+    Ref doc = document();
+    setShouldDelayLoadEvent(doc, true);
```
If the document is already loaded, and we change the src value of an image element, could there be an issue with incrementing/decrementing the counter and triggering another `Document::loadEventDelayTimerFired` call?

annevk replied:
Nothing happens:
```
void DocumentLoader::checkLoadComplete()
{
    if (!m_frame || isLoading())
        return;
```
Earlier versions of this patch would have tested as well as they incremented and decremented for all loads and I didn't encounter issues with that.

youennf commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -486,18 +487,32 @@ void ImageLoader::notifyFinished(CachedResource& resource, const NetworkLoadMetr
+        // It is technically possible state changed underneath us.
+        if (!m_image)
+            return;
+        setShouldDelayLoadEvent(doc, false);
```
What happens for load event delay if `m_image` is null? Or if the image element goes to another document?

annevk replied:
I guess this would be solved by moving it into `subresourcesAreFinished`.

youennf commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -486,18 +487,32 @@ void ImageLoader::notifyFinished(CachedResource& resource, const NetworkLoadMetr
-    loadEventSender().dispatchEventSoon(*this, eventNames().loadEvent);
+    Ref doc = document();
+    setShouldDelayLoadEvent(doc, true);
+    m_image->protectedImage()->subresourcesAreFinished([this, protectedThis = Ref { *this }, doc]() mutable {
```
Maybe `subresourcesAreFinished` should be the one doing the incrementing at call site/decrementing just before calling the completion handler.
This way, no need to add a new `setShouldDelayLoadEvent` virtual method.

Maybe `subresourcesAreFinished` could take a ShouldDelayLoadEvent boolean enum.

annevk replied:
I suppose we could make the document argument optional so it works for all callers of `subresourcesAreFinished`. That might indeed be slightly cleaner. (Note that the document `subresourcesAreFinished` operates on and the one we increment/decrement on are different. `subresourcesAreFinished` is concerned with the SVG image document whereas increment/decrement happens on the document the SVG image document is nested in.)

youennf commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -551,6 +551,18 @@ void SVGImage::subresourcesAreFinished(CompletionHandler<void()>&& completionHan
+    if (m_shouldDelayLoadEvent == shouldDelay)
+        return;
+
+    m_shouldDelayLoadEvent = shouldDelay;
```
It is a bit odd to have this method here, given SVGImage is not able to recompute its document.
I would tend to have the bool in ImageLoader if we need one.

annevk replied:
That would not help with the case I called out in the commit message. Where you have multiple `ImageLoader` instances pointing to the same `SVGImage` instance. Which happens in the various `svg/as-image` tests as they use `svg:use`. But if we merge this into `subresourcesAreFinished()` perhaps the concern goes away?

youennf commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -545,10 +545,16 @@ bool isInSVGImage(const Element* element)
-    internalPage()->localTopDocument()->whenWindowLoadEventOrDestroyed(WTFMove(completionHandler));
+    if (embedderDocument)
+        embedderDocument->incrementLoadEventDelayCount();
+    internalPage()->localTopDocument()->whenWindowLoadEventOrDestroyed([embedderDocument, completionHandler = WTFMove(completionHandler)]() mutable {
```
```suggestion
    internalPage()->localTopDocument()->whenWindowLoadEventOrDestroyed([embedderDocument = WeakPtr { embedderDocument }, completionHandler = WTFMove(completionHandler)]() mutable {
```

youennf commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -545,10 +545,16 @@ bool isInSVGImage(const Element* element)
+    if (embedderDocument)
+        embedderDocument->incrementLoadEventDelayCount();
+    internalPage()->localTopDocument()->whenWindowLoadEventOrDestroyed([embedderDocument, completionHandler = WTFMove(completionHandler)]() mutable {
+        if (embedderDocument)
```
```suggestion
        if (RefPtr document = embedderDocument.get())
```

youennf commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6630,7 +6630,14 @@ void Document::whenWindowLoadEventOrDestroyed(CompletionHandler<void()>&& comple
+        m_whenWindowLoadEventOrDestroyed = WTFMove(completionHandler);
+        return;
+    }
+    m_whenWindowLoadEventOrDestroyed = [oldCompletionHandler = WTFMove(m_whenWindowLoadEventOrDestroyed), newCompletionHandler = WTFMove(completionHandler)]() mutable {
```
```suggestion
    m_whenWindowLoadEventOrDestroyed = [oldCompletionHandler = std::exchange(m_whenWindowLoadEventOrDestroyed, { }), newCompletionHandler = WTFMove(completionHandler)]() mutable {
```

We do not tend to move member variables since they can be reused.

annevk committed:
Improve handling of SVG images with subresources
https://bugs.webkit.org/show_bug.cgi?id=290840

Reviewed by Youenn Fablet.

When a resource is completely fetched, ImageLoader::notifyFinished is
called. This does not quite work for SVG images as they can have local
subresources (data: URLs in particular).

In 292944@main we introduced some infrastructure to handle this better
for ImageUtilities callers. Notably that did not handle multiple layers
of nesting, because it did not improve the situation for HTML <img> and
SVG <image> elements.

In particular if you have a document _DOC_ that references an SVG image
_IMAGE_. And _IMAGE_'s document has an SVG <image> that references
another SVG image _IMAGE2_. And _IMAGE2_'s document references another
image _IMAGE3_, then _IMAGE_'s subresourcesAreFinished() would return
before _IMAGE3_ is ready, because _IMAGE2_ did not increase the delay of
_IMAGE_'s document's load event.

So now we add the infrastructure for those elements to ImageLoader and
SVGImage.

The final bit of complexity here is around supporting multiple
invocations of SVGImage::subresourcesAreFinished by improving
Document::whenWindowLoadEventOrDestroyed. This is needed because
multiple ImageLoader instances can end up pointing to the same SVGImage
instance. E.g., in our test coverage this is hit through SVG <use>
usage.

* LayoutTests/TestExpectations:
* LayoutTests/svg/as-image/svg-image-with-data-uri-from-canvas.html:
* LayoutTests/svg/as-image/svg-image-with-data-uri-reloading.html:
* LayoutTests/svg/as-image/svg-image-with-svg-data-uri.html:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::whenWindowLoadEventOrDestroyed):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::notifyFinished):
(WebCore::ImageLoader::setImageCompleteAndMaybeUpdateRenderer):
* Source/WebCore/loader/ImageLoader.h:
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::subresourcesAreFinished):
* Source/WebCore/platform/graphics/Image.h:
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::subresourcesAreFinished):
(WebCore::SVGImage::tryCreateFromData):
* Source/WebCore/svg/graphics/SVGImage.h:

Canonical link: https://commits.webkit.org/293259@main
